Dimensions: Add Aspect Ratio block support

andrewserong opened the pull request:
<!-- Thanks for contributing to Gutenberg! Please follow the Gutenberg Contributing Guidelines:
https://github.com/WordPress/gutenberg/blob/trunk/CONTRIBUTING.md -->

## What?
<!-- In a few words, what is the PR actually doing? -->

Fixes: https://github.com/WordPress/gutenberg/issues/54032

This PR explores adding aspect ratio to the Cover and Group blocks by including it as part of the dimensions block supports.

It supports aspect ratio at the individual block level, and within global styles. Note that the override for the Cover block that switches off its min-height value only works at the individual block level, and not in global styles.

<!--
A limitation (or feature depending on how you look at it) is that `aspect-ratio` will preserve an aspect ratio so long as there isn't anything else that makes it break out of that aspect ratio. I.e. it'll allow extending _beyond_ a `minHeight`, but not reduce that minimum height. And if the content within the block is too large for the aspect ratio, again the container will extend. In my opinion, I think this is probably pretty decent behaviour, and means we let `aspect-ratio` work as the CSS spec works.

It'll be interesting to see how this feels in real usage, though. -->

## Why?
<!-- Why is this PR necessary? What problem is it solving? Reference any existing previous issue(s) or PR(s), but please add a short summary here, too -->

We currently only have minHeight controls, which doesn't allow for easily creating particular proportions. It'd be helpful for many different kinds of site designs to be able to set container blocks like Cover and Group to a particular aspect ratio.

## How it works (and potential limitations)

This PR tries to let the CSS `aspect-ratio` work the way that it works, with minimal interference. To achieve this, and to avoid overflow issues, when an `aspect-ratio` is set, we `unset` any value for `min-height`. Similarly, if a `min-height` is set at the block level, we output an `aspect-ratio: unset` rule to override any aspect ratio set in global styles. These `unset` rules should only be output for blocks that opt-in to aspect ratio, and are not saved to post content.

For folks wanting to reduce the width of their container to less than the full width of the container, a good way to do it is to wrap the block in a Row block, and use the child width controls, as the flex-basis rules for fixed width appear to play nicely here.

## How?
<!-- How is your PR addressing the issue at hand? What are the implementation details? -->

* Add `aspectRatio` to dimensions block support
* Opt the Cover block into that support (note: Group block support was removed in #58414)
* Add a `min-height: unset` rule that gets applied when aspect ratio is in use — this is necessary because setting a `min-height` interferes with aspect-ratio behaviour by not allowing content to expand the area of a block beyond the defined aspect ratio. In order to ensure that longer content flows correctly in narrower viewports, we unset `min-height` when aspect ratio it applied.
* This PR also uses server-rendering instead of saving the `aspect-ratio` rule into post content. This is because the aspect ratio rule depends on other CSS values, so it's safer if we don't serialize the value for now. This also allows us to explore other options in the future for output, like using `has-4-3-aspect-ratio` classnames that apply the aspect ratio instead of inline styles. And finally, it allows us to add `aspect-ratio` and `unset` rules without having to deprecate any legacy markup.
* Update global styles output to _also_ output a `min-height: unset` rule when outputting `aspect-ratio` — this is a little hacky, but ensures that rules like the Cover block's default `min-height` don't interfere with the aspect ratio.
* Note: for now, the Group block's aspect ratio controls are disabled in global styles, to prevent users from unexpectedly setting aspect ratio for all of the group block's variations

<!--
## To-do

* [x] ~Work out how to safely enable in global styles — currently, you can set an `aspect-ratio`, however we don't also apply the `has-aspect-ratio` classname via global styles, and so the block doesn't receive `width: 100%`, resulting in the block extending beyond the viewport on mobile screen sizes. If the approach in this PR becomes viable, we'll need to either enable this for global styles, or disable in global styles in the shorter-term~ Update: I've hacked in outputting `width: 100%` where needed when outputting the `aspect-ratio` rules.
-->

## Testing Instructions
<!-- Please include step by step instructions on how to test this PR. -->
<!-- 1. Open a post or page. -->
<!-- 2. Insert a heading block. -->
<!-- 3. etc. -->

1. Add Cover and Group blocks to a post or page.
2. Set background images and colors on the blocks.
3. Try selecting from a range of aspect ratios under the Dimensions panel.
4. Try playing around with adding `minHeight`, too — notice that `minHeight` will clear out any aspect ratio applied.
5. To reduce the overall size of the block, try wrapping in a Row block, and use the Fixed width controls to reduce the size of the block.
6. Also, try playing around with aspect ratio at the block level in global styles.

<!--
6. Note: When the block is the child of the _Stack_ block, aspect ratio doesn't really do anything when adjusting the fixed height, as the fixed height will always win. -->

In testing, how do these controls feel to use? Is it close enough to desired behaviour to be useful, or does it feel limiting? Try adding a variety of content within the Group and Cover blocks to get a feel for it, too.

## Screenshots or screencast <!-- if applicable -->

https://github.com/WordPress/gutenberg/assets/14988353/26d749c1-6bd5-403a-b844-c323db8cf644

<!--
https://github.com/WordPress/gutenberg/assets/14988353/a3fc3b5f-eca1-4632-b868-ee78cbe6aa27
-->

andrewserong committed:
Try: Dimensions Aspect Ratio block support

andrewserong committed:
Try server-rendering output

andrewserong committed:
Try outputting has-aspect-ratio classname

andrewserong commented:
Note to self: for some reason the `has-aspect-ratio` classname only appears to be output when other things like border color is set... weirdly it doesn't appear to be enough for it to update the classname in the editor on its own. I'll dig into this further.

Also, I had to add a `width: 100%` rule for the aspect ratio to not break out of its container, and for it to respect flex basis rules. I'll do some more testing on that, too.

andrewserong commented on packages/block-editor/src/hooks/dimensions.js:
```diff
@@ -131,6 +135,48 @@ function DimensionsPanelPure( { clientId, name, setAttributes, settings } ) {
+		'has-aspect-ratio': !! style?.dimensions?.aspectRatio,
+	} );
+
+	return { className };
```
TODO: Add more checks before output here. I've left this pretty bare for the moment while I debug why `has-aspect-ratio` is only output some of the time 🤔

andrewserong commented on packages/block-editor/src/components/block-list/block.js:
```diff
@@ -53,10 +53,18 @@ function mergeWrapperProps( propsA, propsB ) {
 
-	if ( propsA?.className && propsB?.className ) {
+	// May be set to undefined, so check if the property is set!
+	if (
+		propsA?.hasOwnProperty( 'className' ) &&
+		propsB?.hasOwnProperty( 'className' )
+	) {
 		newProps.className = classnames( propsA.className, propsB.className );
 	}
-	if ( propsA?.style && propsB?.style ) {
+
+	if (
+		propsA?.hasOwnProperty( 'style' ) &&
+		propsB?.hasOwnProperty( 'style' )
+	) {
```
This merges in a fix that will land in https://github.com/WordPress/gutenberg/pull/56912 to ensure classnames are output correctly. Just including it here for now so that this PR will work properly in manual testing 🙂

andrewserong commented on packages/block-library/src/cover/style.scss:
```diff
@@ -17,6 +16,16 @@
+		min-height: 430px;
+	}
+
+	&:where(.has-aspect-ratio) {
+		width: 100%;
+	}
```
We'll likely need to put this rule somewhere else as I think we'll always want this applied when `.has-aspect-ratio` is active.

andrewserong committed:
Ensure block support is checked properly

ramonjd commented on lib/block-supports/dimensions.php:
```diff
@@ -74,6 +74,65 @@ function gutenberg_apply_dimensions_support( $block_type, $block_attributes ) {
+
+			if ( ! empty( $styles['classnames'] ) ) {
+				foreach ( explode( ' ', $styles['classnames'] ) as $class_name ) {
+					$tags->add_class( $class_name );
```
Mainly interested... so `$tags->add_class( 'a-class b-class' )` should work too?

andrewserong replied:
At the moment I'm calling `explode` so that we add them one-by-one... I looked up the `add_class` method and it does look like we should be able to call it with a full classname string that contains multiple classes, I just wasn't sure if it's _guaranteed_ to 🤔. Happy to switch it over to the simpler `add_class` with the full single string, though 🙂

Thanks for giving this PR an early look! 🙇

ramonjd replied:
I gave it a try and it works 😄 No biggie, I was just wondering.
<img width="650" alt="Screenshot 2023-12-12 at 4 43 30 pm" src="https://github.com/WordPress/gutenberg/assets/6458278/b3001712-9ac2-45e1-beb0-1b147e9a37f9">

andrewserong replied:
Nice one! I can switch it over 😀

tellthemachines replied:
Yeah if it's not explicitly indicated in the docs it might not be officially supported even though it currently works. We were using similar logic to fetch a bunch of classes from an element in one go in the layout support, but [this change](https://github.com/WordPress/wordpress-develop/pull/5096) broke that ability. The refactor of layout in #54075 to fix that has more info.

ramonjd replied:
Agree - might be a nice feature though to have an `add_classes` method that accepts an array or something.

andrewserong replied:
Thanks folks, good points! Related to what you've linked to in #54075, I just did a quick test of what happens if the class string includes multiple classes, and it seems it gets treated as a single classname so the de-duping logic in `add_class` gets skipped, resulting in potential duplicates:

<img width="667" alt="image" src="https://github.com/WordPress/gutenberg/assets/14988353/0a763e53-a53d-4884-84a1-bcf5ccfe188b">

I'll stick with the `foreach` for now 🙂

ramonjd commented:
Awesome stuff! I know it's early days, but it's working a treat so far. Looking forward to testing it again


https://github.com/WordPress/gutenberg/assets/6458278/174efecc-3130-4c03-94a9-21f113de55bc

jasmussen commented:
This is fantastic. Functionally it is working exactly as I had hoped/expected. Here's a GIF showing the new aspect ratio control, with frontend preview, fully functional:

![cover](https://github.com/WordPress/gutenberg/assets/1204802/73b858b1-ceb8-4f96-9887-af730d8f9804)

The main question is where to put this control, where in this PR you've put it in the Dimensions panel:

![Screenshot 2023-12-19 at 10 34 50](https://github.com/WordPress/gutenberg/assets/1204802/e9b90545-d9b9-4bc6-99c9-3fa60ce3bccf)

This is arguably the correct way to place it, and follows a pattern established by the Featured Image block as well:

![Screenshot 2023-12-19 at 10 35 22](https://github.com/WordPress/gutenberg/assets/1204802/2a266198-4af8-4a19-9e01-c39f6a28c2d3)

The odd one out is the Image block, which puts it in Settings:

![Screenshot 2023-12-19 at 10 33 53](https://github.com/WordPress/gutenberg/assets/1204802/31b62fe9-89d4-4549-8ce6-2c78e1a3dd05)

A soft ping to @WordPress/gutenberg-design for input here; what do we see as the unified end-state for this control? Would it make sense to (separately) move the dimensions controls from the Image settings tab to the style tab, and what would then happen to the Alt text? 

On a separate note, I would expect the Cover block to eventually also get _width_ controls on every block (#25973), so in principle we can move towars this design for the Cover block height control even today:

![Screenshot 2023-12-19 at 10 38 41](https://github.com/WordPress/gutenberg/assets/1204802/32f788e9-e474-46f5-82ab-00f874a59dca)

We can just call it "Max height" instead of "Height". "Maximum height of Cover" is unusually verbose and slightly redundant. Not a blocker here, but simply thinking ahead.

Nice work, thank you!

jameskoster commented:
> I would expect the Cover block to eventually also get width controls on every block

That feels like an important consideration. There's already something a little awkward about the contradiction that can occur between min-height and aspect ratio. If you can define static height / width, how does the interplay with aspect ratio work?

For this PR specifically, is intended that content becomes invisible if it is too 'tall' for the container? Should there be `overflow: auto` or similar?

jasmussen commented:
> If you can define static height / width, how does the interplay with aspect ratio work?

If you have both an aspect ratio, and a min-height, the aspect ratio will define the height until it reaches the min-height, which is what I'd expect. 

> For this PR specifically, is it intended that content becomes invisible if it is too 'tall' for the container? Should there be overflow: auto or similar?

What are you seeing here?

jameskoster commented:
> If you have both an aspect ratio, and a min-height, the aspect ratio will define the height until it reaches the min-height, which is what I'd expect.

That's not how it works though? If I set aspect ratio to 1:1 _and_ min-height to 1000000px then the aspect ratio will never be respected because there's not enough horizontal space to do so.

> What are you seeing here?

If the height of the contents inside the Cover greater than the aspect ratio allows for the Cover height then the content gets cut off at the top and bottom. Hopefully this video explains better:


https://github.com/WordPress/gutenberg/assets/846565/33c0efe7-877d-457f-8339-5ff6b7a82d2b

jasmussen commented:
> That's not how it works though? If I set aspect ratio to 1:1 and min-height to 1000000px then the aspect ratio will never be respected because there's not enough horizontal space to do so.

Right, apologies, I miswrote my sentence, I meant "reaches" in the scaling down kind of way. A different way to say it is: min-height is always min-height, so if the height as inferred by the aspect ratio is smaller than the min-height, the min-height will serve s the floor value. Which is what I'd expect, it remains the _minimum_ value.

> If the height of the contents inside the Cover greater than the aspect ratio allows for the Cover height then the content gets cut off at the top and bottom. Hopefully this video explains better:

Right, yes, thank you for clarifying.

Honestly I think we can explore an overflow design tool as a separate thing, and leave it to an editorial responsibility to ensure things look good here.

andrewserong committed:
Try adding global styles support

andrewserong committed:
Update docs, add appropriate checks

andrewserong commented:
Thanks for the reviews, testing, and feedback folks! Glad to hear the overall direction sounds promising to everyone. I've tidied up this PR a little and added in global styles support. It's a _little_ hacky as it'll need to output the `width: 100%` rule in order for things to play nicely, but let's see how it looks for everyone during review. Since it sounds like we're on the home stretch with this one, I've moved it out of draft, so it's ready for revew!

> On a separate note, I would expect the Cover block to eventually also get width controls on every block (https://github.com/WordPress/gutenberg/issues/25973), so in principle we can move toward this design for the Cover block height control even today:

Good to know! As part of this PR I've added a tiny bit of support to the style engine so that it can handle `width` values, which should prep us a little bit for eventually adding in proper support for it.

andrewserong commented on packages/style-engine/class-wp-style-engine.php:
```diff
@@ -182,6 +191,12 @@ final class WP_Style_Engine {
 					'spacing' => '--wp--preset--spacing--$slug',
 				),
 			),
+			'width'       => array(
+				'property_keys' => array(
+					'default' => 'width',
+				),
+				'path'          => array( 'dimensions', 'width' ),
+			),
```
This is just to support outputting `width` values for now — this PR doesn't propose adding a width block support just yet, but includes it in the structure of the `style` object for now so that it's a little tidier for outputting the `width: 100%` rule alongside the `aspect-ratio` rules.

tellthemachines reviewed (commented):
Interesting problem!

> Add a width: 100% rule that gets applied when aspect ratio is in use — this is necessary because as soon as an aspect-ratio rule is applied, the block will fill its container to its max-width, which will extend the block outside of the width of the viewport on mobile screen sizes.

Under what specific circumstances does the block extend outside viewport width? It sounds like a bug, not the expected behaviour of a block extending to its container width 🤔 

I'm asking this mostly because the `width: 100%` rule breaks full-aligned containers with negative padding:

<img width="638" alt="Screenshot 2023-12-20 at 4 22 18 pm" src="https://github.com/WordPress/gutenberg/assets/8096000/b5a0efe6-08fa-43f7-868c-f117bde16c94">

and I couldn't reproduce the horizontal scroll just playing around with different group widths and viewport sizes.

andrewserong commented:
> Under what specific circumstances does the block extend outside viewport width? It sounds like a bug, not the expected behaviour of a block extending to its container width 🤔
> I'm asking this mostly because the width: 100% rule breaks full-aligned containers with negative padding:

Great catch! That's an unfortunate problem, indeed! 😅

Now that I'm re-testing it without the `width: 100%` rule, I'm not able to reproduce the issue I originally ran into. Leave it with me, and I'll give this a proper re-test with the `width: 100%` rule removed tomorrow and see what the cause was... it'd be ideal if we can get away with not introducing that rule.

Thanks for giving this a spin!

andrewserong commented:
Update: I remember now... unfortunately the `width: 100%` rule is required because of what happens if we set a minimum height to a block. As soon as there's a minimum height set, the `aspect-ratio` will cause the element to overflow horizontally 😞

This might be tricky to come up with a fix for 🤔

andrewserong commented:
Another potential wrinkle is that the presence of setting `minHeight` appears to cause `aspect-ratio` to treat it as `height` and so content will overflow beyond it. With the Cover block, that content gets clipped. But the Group block, it overflows awkwardly (note the end of the green area, but paragraph content extends past the next heading block):

<img width="1395" alt="image" src="https://github.com/WordPress/gutenberg/assets/14988353/9b5f1395-0219-4ade-8c73-2159cada9b5c">

So it is possible that we'll wind up causing friction for users here 🤔

I think that leaves the potential blockers discovered for now as:

* `width: 100%` breaks full-aligned containers with negative padding
* `minHeight` being set on the block causes longer content to overflow (Group block) or be clipped (Cover block)

jasmussen commented:
> Update: I remember now... unfortunately the width: 100% rule is required because of what happens if we set a minimum height to a block. As soon as there's a minimum height set, the aspect-ratio will cause the element to overflow horizontally 😞

Would a _max-width_ fix the issue? That works in [this codepen](https://codepen.io/joen/pen/WNmeEgx?editors=1100), which is 16:9 in most cases, but it has a tall min-height so it _should_ break out of the main column if not for the max-width.

andrewserong commented:
> Would a max-width fix the issue?

Unfortunately no. It kind of technically would, except that it then interferes with the content size of whichever container the block is in, and then the block stretches all the way out:

![image](https://github.com/WordPress/gutenberg/assets/14988353/1c33b22c-14b1-4131-82d7-a302389d0c64)

It's a little frustrating as I've had a bunch of fun playing around with `aspect-ratio` in Codepen, too, and it looks _so_ close to just working out of the box! I've looked through the other options for `width`-based values, but haven't had much luck unfortunately, and we always wind up with issues for the full-width option when root padding is in use 🤔.

I'm a little stumped by this one, so I might need to set this PR aside over the holidays and come back to it fresh in the new year.

Thanks again for the feedback and discussion, folks!

tellthemachines commented:
> * `width: 100%` breaks full-aligned containers with negative padding
> * `minHeight` being set on the block causes longer content to overflow (Group block) or be clipped (Cover block)

This is sounding very much like the min height and aspect ratio settings are incompatible. I think conceptually that makes sense: when setting an aspect ratio I'd expect that to be respected at all screen sizes, and a min height will interfere with it.

Can we consider disabling min height when aspect ratio is set? Perhaps with a warning to users when they try to enable aspect ratio on a block that already has min height? It's great when things just work as if by magic, but not everything is possible 🤷

jasmussen commented:
> Can we consider disabling min height when aspect ratio is set? 

Good idea! We can, yes, that might be a good solution. They are two tools to accomplish a height, after all.

After all, for the image block the width and height controls show "Auto" unless set:
![image auto](https://github.com/WordPress/gutenberg/assets/1204802/96829f00-6ec8-4472-abaa-513df0cdbbbd)

We can do something similar here:

* Default/initial state, no aspect ratio is set, the dropdown says "None"
* If the min-height is further set, the dropdown continues to say "None"
* If you set an aspect ratio, the min-height field is cleared, and now says "Auto"

Same as image. You choose whether you want explicit dimensions, _or_ an aspect ratio.

jameskoster commented:
My main concern was that having active width/height/min-height _and_ aspect ratio settings felt broken – it wasn't clear which took priority. The suggestion above would address that 👍

andrewserong commented:
Thanks for the feedback and ideas, folks!

I really like the idea of removing the `min-height` value when an aspect-ratio is set. I think this means, though, that we'd need to restrict this block support to only be available at the individual block level, and not in global styles. Does that sound like a blocker to anyone? Otherwise it might be hard to figure out the right order of things so that a user can set a minimum height at the individual block level, overriding global styles' aspect-ratio and switching that off, etc.

From the sounds of it, I imagine the main use case for aspect-ratio is for individual blocks and patterns rather than "make all group blocks 4:3" or "make all cover blocks 16:9", so just enabling at the block level sounds reasonable to me, but just thought I'd check before digging in.

I'll likely run out of time today to get much further with this, but will continue hacking on it in the new year! Overall, this is sounding promising, though 😄

ajlende commented:
> This means that if a block (e.g. Cover or Group) has a `minHeight` that is greater than the height of the `aspect-ratio` calculation, then the `minHeight` rule wins. This has the benefit of meaning that folks can set a minimum height that isn't the selected aspect ratio so that things still look good on mobile, and then as the viewport scales up, the aspect ratio is preserved.

Having seen some support tickets with confusion around mobile. The mindset that most users have is that mobile and desktop are completely different things. This means that when a user sets an aspect ratio, the general expectation is that aspect ratio to be true in both mobile and desktop. It's unfortunate that CSS has different opinions about how it should work because it makes it harder to build in `aspect-ratio` to Gutenberg.

> I really like the idea of removing the `min-height` value when an `aspect-ratio` is set.

As others have said, this is the solution that I ended up for with the image block. When you're setting/unsetting multiple values at once, I found it helpful to diagram out all the interactions to ensure that none were missed. See https://github.com/WordPress/gutenberg/pull/51545#issuecomment-1601326289 for the full set of diagrams.

andrewserong commented:
> When you're setting/unsetting multiple values at once, I found it helpful to diagram out all the interactions to ensure that none were missed. See https://github.com/WordPress/gutenberg/pull/51545#issuecomment-1601326289 for the full set of diagrams.

Wow, that's awesome, thanks for the link and for your insights into the prior work here!

I'll have another go at all this in the new year 🙂

jasmussen commented:
>  I think this means, though, that we'd need to restrict this block support to only be available at the individual block level, and not in global styles. Does that sound like a blocker to anyone? Otherwise it might be hard to figure out the right order of things so that a user can set a minimum height at the individual block level, overriding global styles' aspect-ratio and switching that off, etc.

Hmm, that's a doozy. I actually could see myself setting all Cover blocks to be 16:9, mainly because for some use cases I would see the Cover block _not_ as a fancy Group block with innerblocks and backgrounds, but as an Image but with a text overlay. Thinking of Image and Cover together, I admit, is going to vary a bit from user to user and even theme to theme, but nevertheless feels valid to me.

Would it be possible to add the global aspect ratio support for just Cover, but not Group?

Or: what's the main motivation for not wanting to set this control as available globally?

andrewserong commented:
Good to know there's a use case for global styles 👍

> Or: what's the main motivation for not wanting to set this control as available globally?

The main problem is that if aspect ratio is set at the block level in global styles, and a user then goes to add a minimum height at the individual block level in post content, then we're back at the problem of a `min-height` rules interfering with `aspect-ratio` output.

Perhaps a workaround for that case could be to output `aspect-ratio: unset` at the individual block level when applying a `min-height` value so that setting a minimum height explicitly switches off the aspect ratio? That gives me something else to hack around with. (In this example, we have set Cover globally to an aspect ratio, and then want a single cover block to use min-height instead).

For now, though, let's accept global styles support as a requirement since you've provided a good use case for it, and we can see what blockers we run into in unsetting `min-height`s when we're intending to output an aspect ratio.

Let me know if anyone has other ideas about how to resolve the conflicts, though, but I think this gives us something to go on with for now 🤞

tellthemachines commented:
> The main problem is that if aspect ratio is set at the block level in global styles, and a user then goes to add a minimum height at the individual block level in post content, then we're back at the problem of a `min-height` rules interfering with `aspect-ratio` output.

If we have access to a block's global style settings at the individual block level, we can still disable min height when aspect ratio is set globally. Then unsetting aspect ratio at the block level could be a requirement for setting min height (we could show a warning or something)

> Perhaps a workaround for that case could be to output aspect-ratio: unset at the individual block level when applying a min-height value so that setting a minimum height explicitly switches off the aspect ratio?

I like the idea of being able to set min height without manually switching off aspect ratio first, but it will be confusing unless we also toggle aspect ratio off. Perhaps we could do that and add a warning that enabling min height will disable aspect ratio? We need to somehow make it clear to users that these settings are mutually incompatible.

jameskoster commented:
> We need to somehow make it clear to users that these settings are mutually incompatible.

I have a hunch that automatically setting aspect-ratio to `unset` might be adequate, and would lean towards trying that first. In most cases I suspect the incompatibility will seem kind of obvious. If it turns out to _not_ be obvious then we can look at adding some friction to proceedings.

Might we try the inverse too, IE if min-height is set globally and you add an aspect ratio locally then min-height is set to `none` locally?

andrewserong committed:
Try unsetting minHeight

andrewserong commented:
Tiny update: I've switched this over to Draft while I'm still hacking around on it. Will ping again when it's ready for another review.

For now, though, quickly hacking in `unset` for `min-height` when outputting `aspect-ratio` resolves the issue of content overflow for me, so seems like a promising direction overall:

| 16:9 aspect ratio without much content | When more content is added the block naturally extends now |
| --- | --- |
| ![image](https://github.com/WordPress/gutenberg/assets/14988353/caf0be0d-d7f9-41bc-b031-5a1f0836cfd8) | ![image](https://github.com/WordPress/gutenberg/assets/14988353/0c0a30b7-d286-4993-a2a8-f99c129d50f4) |

Next up, I'll have a play with how we output rules in global styles, and what to do when we've set `minHeight` at the individual block level in post content, and what to do about the output there (it might be enough to simply output an `aspect-ratio: unset` if we have a `min-height` value, as an inverse of the logic I just added to the `aspect-ratio` support. I'll have a play.

andrewserong committed:
Try unsetting min-height when applying aspect ratio and vice versa

andrewserong committed:
Allow Cover block to expand when the content extends beyond the boundaries of the aspect-ratio

andrewserong committed:
Hide cover block resize handle when aspect ratio is set, clear aspect ratio when updating cover block min height

andrewserong commented:
Update: I've pushed a few changes that gets this working a little more reliably:

* Allow the Cover block to expand when content extends beyond the boundary of the aspect-ratio by swapping out the `overflow:clip` and fallback `overflow: hidden` rules that are used to ensure border-radius clips correctly with `overflow-x: clip` — this keeps the clipping, but ensures that the `y` dimension can expand based on how much content there is. So, for example, you can set a Cover block to 16:9, but then in narrow viewports, instead of the content clipping, it'll now extend the Cover block down vertically so that text isn't cut off.
* Hide the cover block resize handle when an aspect ratio is applied, otherwise the drag handle obscures the block.
* Clear out local `aspect-ratio` value when updating the Cover block's ad hoc `minHeight` value
* The minHeight and aspect ratio block supports now each clear out the other one when you go to make a change on the Group block, which seems to work nicely:

https://github.com/WordPress/gutenberg/assets/14988353/a497a235-25aa-4c4c-89a6-733558509a71

---

The main issue I'm running into now is with trying to get things working nicely with global styles output without having to add too much workaround logic / without blocks having to know too much about the aspect ratio support. Here's an example:

* Update the Cover block to globally use a 16:9 aspect-ratio
* At the block level for the Cover block, give it a min-height
* The Cover block doesn't know that there is a global style for it that gives it an aspect ratio, so it doesn't know to set `aspect-ratio: unset` in its output to counteract the global style rule for aspect ratio. So, we get the old overflow issue appearing again:

https://github.com/WordPress/gutenberg/assets/14988353/a18a3e35-8677-4cfc-a67f-ae3787e133a2

So, I think the trick here will be to come up with a way to elegantly output `aspect-ratio: unset` when a `min-height` rule is applied locally to a block. I _think_ I have an idea for how to do that in the PHP output (dynamic). Something like:

* If the block supports aspect ratio, and a min height value is in use (either from the block support or from the Cover block's ad hoc control) then, output `aspect-ratio: unset`. This is the inverse of what I'm currently doing by setting `min-height: unset` when an aspect ratio is set. I'm pretty sure this should work okay.
* This approach doesn't require looking up whether an aspect-ratio value is _really_ set, it can just know to output the `unset` when a `min-height` value is set, and everything should be fine.
* The tricky part is in the block editor, where I'll need to dynamically inject an unset based on what's in block attributes — this is where I'm a little stuck at the moment, as we currently expect to handle a `style` object as opposed to output extra rules based on random other attributes (in this case `minHeight` from the Cover block).

---

TL;DR — I think I've handled most of the cases of setting aspect-ratio and clearing out min-height, but there's still the aspect-ratio set at global styles issue to contend with. Overall, I _think_ this is feeling fairly promising. I'll continue hacking away at the global styles problem, but just wanted to give an update on where I'm up to 🙂

andrewserong committed:
Clear aspect-ratio if minHeight is set on the site frontend

andrewserong commented:
Update to the update:

> If the block supports aspect ratio, and a min height value is in use (either from the block support or from the Cover block's ad hoc control) then, output aspect-ratio: unset. This is the inverse of what I'm currently doing by setting min-height: unset when an aspect ratio is set. I'm pretty sure this should work okay.

Turns out the PHP approach seemed to be pretty straightforward, implemented in: https://github.com/WordPress/gutenberg/pull/56897/commits/ef7149df4dd95dcb33c80b1ffe3d95bac55f5622

Now to figure out how to do basically the same thing in JS without serializing to post content (since we only want to be outputting these extra rules dynamically, not baking them in to post content).

jasmussen commented:
> So, I think the trick here will be to come up with a way to elegantly output aspect-ratio: unset when a min-height rule is applied locally to a block. I think I have an idea for how to do that in the PHP output (dynamic).

What problems would it introduce if we just always output `aspect-ratio: none;` together with a min-height?

andrewserong commented:
> What problems would it introduce if we just always output aspect-ratio: none; together with a min-height?

So far, it doesn't look like it causes any problems that I can tell! 🤞
It seems to be working nicely on the PHP side, I just now need to apply similar logic in the JS side. There's a couple of reasons that it's good to apply this dynamically rather than saving into post content:

* So that things still work nicely for older post content without folks needing to update markup
* So that we don't need to add any deprecations
* And so that we can change the output rules further down the track if and when we add additional block supports that might interact with aspect ratio

I'll dig into the JS logic a bit further and see what we can come up with.

andrewserong committed:
Attempt to get unset rules playing nicely in the editor with global styles

andrewserong commented:
Update: I think I've managed to get it working pretty well in https://github.com/WordPress/gutenberg/pull/56897/commits/2c9107d2ff9a04b660053b5cf28fb1e7d04c5247

And so far, that seems to enable setting the Cover block to an aspect ratio in global styles while still allowing users to override with a manual min height at the block level in the post editor along with swapping to a different aspect ratio:

<img width="1294" alt="image" src="https://github.com/WordPress/gutenberg/assets/14988353/195920f4-6019-4d7e-9348-7f8aea33a59f">

Also, the rules now seem to (mostly) reliably allow content that extends beyond the bounds of the aspect ratio to expand the block, which means that things _should_ work pretty well across different viewport sizes.

The main situation I encountered where this breaks down a bit is if folks attempt to set an aspect ratio in global styles for the Group block and then use a Stack block where it's set to wrap items. This winds up resulting in this odd situation where the items in the Stack _don't_ wrap as expected and things overflow:

![image](https://github.com/WordPress/gutenberg/assets/14988353/87b8d3e0-d5eb-4dca-ac16-761681e7c6c1)

Ironically, the solution for patterns that have that kind of layout (like in the footer of the TT4 theme) is to set the Stack block to _not_ allow wrap to multiple lines which, counterintuitively, gets things wrapping correctly:

<img width="1278" alt="image" src="https://github.com/WordPress/gutenberg/assets/14988353/71ffdd28-44bb-45f5-80ff-e7a43775ce09">

As that gets quite complex, I don't think it's a problem we should try to fix in this PR, but it's something to be aware of if folks wind up setting the Group block to use an aspect ratio. And it might be a factor in whether or not we think this feature should be exposed (for the Group block at least) in global styles 🤔

---

All up, I _think_ this PR is getting closer to a viable state now. I'll leave it as Draft as I've only _just_ gotten it to a place where I think it's working pretty well. I'll do some more testing next week, but happy for any feedback if folks get a chance to play with this PR again.

jasmussen commented:
Thansk for working on this. Let folks know when this is in a place to test more deeply, it'll be a nice one!

andrewserong commented:
Update: after manually testing again, I think this PR is in a good place for deeper testing, now! The current behaviour strikes (hopefully) a pretty decent balance between trying to output the CSS as-is, while unsetting conflicting `min-height` but otherwise not being too intrusive (since we're using dynamic rendering instead of saving rules to post content). A few highlights:

* Cover block works nicely with aspect ratio set in global styles
* Global styles aspect ratio will be overridden by local changes to min height or aspect ratio
* Updating aspect ratio or min height locally will result in the last changed value being the "real" one / the one that wins, with the other being cleared out, or not prioritised

Let me know if anyone runs into any show-stoppers!

paaljoachim commented:
I am trying this PR at http://gutenberg.run/56897 but I am getting the message "Invalid pull request"
so I went with this instead: https://make.wordpress.org/design/2021/03/03/testing-a-gutenberg-pull-request-pr/

Testing.
Cover block. Image with a border and radius. Switched between various aspect ratio options. This worked fine on backend and frontend.

Group block. 
Testing.
Added a minimum height of 600px to show the background image. Image with a border and radius. Switched between various aspect ratio options. This worked fine on backend and frontend.

From my testing the Cover and Group block image background worked fine!

------

Brain storming while I was testing.

Group block. Background opens to show Solid and Gradient tabs. Should image be included here? As another tab? Because it is part of the Background along with Solid and Gradient.

Should the post/page featured image feature option variation be reused in the Background image option? Showing the image and adding in the Focal Point Picker. Gathering together the options in the Background Image button drop down and the focal picker in one area.

andrewserong commented:
Thanks for testing, Paal!

> Should the post/page featured image feature option variation be reused in the Background image option? Showing the image and adding in the Focal Point Picker. Gathering together the options in the Background Image button drop down and the focal picker in one area.

Great ideas, I believe these are all covered in the background image block support tracking issue (https://github.com/WordPress/gutenberg/issues/54336) so the long-term intention is to include featured image support, focal picker, and grouping controls together there to be more discoverable.

richtabor commented:
Seems to work well, but like @jasmussen mentioned a bit ago, it does get lost in this placement. It's consistent with the featured image block, but not image. Perhaps aspect ratio is its own panel, or perhaps we can align on a "Media" panel of sorts, like the site logo block? 

![CleanShot 2024-01-23 at 17 54 30](https://github.com/WordPress/gutenberg/assets/1813435/7087c73a-a203-4bbe-8b25-83899dd97d9c)

andrewserong commented:
> Perhaps aspect ratio is its own panel, or perhaps we can align on a "Media" panel of sorts, like the site logo block?

That is a good question! To me, I was thinking that aspect ratio best belongs alongside other dimensions controls since both aspect ratio and minimum height control the apparent height of the block 🤔

tellthemachines commented on lib/class-wp-theme-json-gutenberg.php:
```diff
@@ -2093,6 +2097,15 @@ protected static function compute_style_properties( $styles, $settings = array()
+					'value' => 'unset',
+				);
+			}
+
```
Looks like min-height is already getting removed when aspect-ratio is set in global styles, so do we still need this here?

andrewserong replied:
Yes — we need the `min-height: unset` in global styles output (which is this part of the code) so that we unset the Cover block (or any other block's) default min-height value that might be in use.

tellthemachines commented on packages/block-editor/src/components/global-styles/use-global-styles-output.js:
```diff
@@ -434,6 +434,12 @@ export function getStylesDeclarations(
+		if ( cssProperty === 'aspect-ratio' ) {
+			output.push( 'min-height: unset' );
+		}
+
```
Same as above, do we need this here, given that we're already deleting min-height when setting aspect-ratio?

andrewserong replied:
Yes, if a block only has aspect-ratio set in global styles and not at the individual block level, we still need to unset any `min-height` rule that might be coming from the block's CSS (as is the case with the Cover block's default CSS)

tellthemachines replied:
Oooh I see, this is intended to cover legacy min-height implementations. Thanks for explaining!

tellthemachines commented on packages/style-engine/class-wp-style-engine.php:
```diff
@@ -185,7 +185,16 @@ final class WP_Style_Engine {
+				),
+				'path'          => array( 'dimensions', 'aspectRatio' ),
+				'classnames'    => array(
+					'has-aspect-ratio' => true,
```
We might need to check the actual value of the aspect-ratio property before applying the class, because it's being applied even when aspect-ratio is `unset`:

<img width="653" alt="Screenshot 2024-01-24 at 10 47 03 am" src="https://github.com/WordPress/gutenberg/assets/8096000/ba2194ca-223b-4f18-86b4-dc666759142e">

andrewserong replied:
Oh, good catch! I'll look into it 👀

andrewserong replied:
There might be a better way to do it, but I've gone with skipping the classname output in `dimensions.php` if the value wasn't originally set in https://github.com/WordPress/gutenberg/pull/56897/commits/7e14457ce2ec28ed2cbd8756ca85c17051284d5c

I _think_ that might be preferable to handling it in the style engine as it'd probably be a bit of a larger change to get the style engine to accept a callback to determine whether or not to output the classname, and what we're doing with the aspect ratio support is a bit different to all the other block supports. If we find we have another use case for this kind of skipping logic, we could always look at updating the style engine in a follow-up?

The change appears to be working okay for me now:

<img width="732" alt="image" src="https://github.com/WordPress/gutenberg/assets/14988353/33135146-adc0-4dbc-99b3-c94f3665914b">

Let me know if there's another way you think we should try it, though! 🙂

tellthemachines reviewed (commented):
This is looking pretty good and testing really well for me! Just a few comments below.

> The main situation I encountered where this breaks down a bit is if folks attempt to set an aspect ratio in global styles for the Group block and then use a Stack block where it's set to wrap items. This winds up resulting in this odd situation where the items in the Stack don't wrap as expected and things overflow

I'm inclined to consider this a feature for the Stack block, because it now allows flex-wrapping to flow the content into columns:
<img width="640" alt="Screenshot 2024-01-24 at 10 39 16 am" src="https://github.com/WordPress/gutenberg/assets/8096000/6fb113d9-2e59-4652-886a-83e69cc08f8a">
I do think it's very awkward to set aspect-ratio globally on all variations of Group block though.

> It's consistent with the featured image block, but not image. Perhaps aspect ratio is its own panel, or perhaps we can align on a "Media" panel of sorts, like the site logo block?

I'm not sure "media" is appropriate given aspect-ratio can now be used on blocks that don't necessarily have any image. I agree there's some consolidation work to do in the inspector tools, but it might be best to leave that for a follow-up task, particularly as it isn't 100% clear what that should look like yet.

andrewserong committed:
Hide aspectRatio control from global styles for the group block

andrewserong commented:
Thanks for reviewing @tellthemachines!

> I do think it's very awkward to set aspect-ratio globally on all variations of Group block though.

Agreed. Since this is a problem to do with the Group block specifically and the way we use variations for it (i.e. we can't allow users to set an aspect ratio _per_ Row / Stack variation), I've had a go at hacking in an exception so that we hide the aspect ratio controls in global styles specifically for the group block, with an explanatory code comment in https://github.com/WordPress/gutenberg/pull/56897/commits/39873bf9e2ec0e32194d94b76b5e5e7cc0550bd6. Let me know if anyone can think of a better way of handling it. I'd rather not try to build an API around it for now, though, as it really is an exception for the Group block, so I'm hoping we can go with something simple 🤞

andrewserong committed:
Skip output of has-aspect-ratio if value isn't set

richtabor commented:
> To me, I was thinking that aspect ratio best belongs alongside other dimensions controls since both aspect ratio and minimum height control the apparent height of the block 🤔

It makes padding and margin feel out of place then. Are those dimensions? :sweater

richtabor commented:
Apologies for the close/open — GitHub UI jumped on me!

noisysocks commented:
Took this for a spin! Tested setting aspect ratios on Groups and Covers, in the block and in global styles, in the post editor and in the site editor.

All worked really well—nice job. Here's the few things I ran into:

- In the Cover block, when I set a minimum height _and then_ an aspect ratio, the cover no longer obeys the minimum height. I'd expect it to reset the minimum height when I do this, which is what happens in the Group block.

  https://github.com/WordPress/gutenberg/assets/612155/25ec4bd8-1372-486f-b30e-de2abd60c39f

- It makes sense to me that you can control the default Aspect ratio for all Cover blocks using Global Styles, but I'm not sure about Group? Group is heavily used by block themes for layouts and so setting it nearly always makes the entire site look broken.

- The aspect ratio dropdown displays _Original_ as the selected item both when there is no aspect ratio set on the block **and** when the aspect ratio is explicitly set to `auto`. I find this a bit confusing—it made me think that global styles inheritance wasn't working.

  https://github.com/WordPress/gutenberg/assets/612155/c43a079d-0a90-40cd-9379-8a2ff65cbdd4

richtabor commented:
> but I'm not sure about Group? Group is heavily used by block themes for layouts and so setting it nearly always makes the entire site look broken.

I agree. It feels a bit out of place to apply to a group block. 

> I found it a bit odd that Featured Image has an Aspect ratio control but that it a) didn't display the same values as Cover and Group; and b) wasn't adjustable using Global Styles.

The post featured image block's implementation is dated and needs to be updated to better reflect this (and the image block's) approach. Every aspect ratio implementation should be the same. Related https://github.com/WordPress/gutenberg/issues/47962.

noisysocks reviewed (commented):
Code looks good 👍

andrewserong commented:
> It makes padding and margin feel out of place then. Are those dimensions? 😅

It's a good question. This might be subjective, but to me having all the controls that relate to spacing and dimensions in the one place is a good thing as it consolidates all the controls that affect the sizing of the element's box. We also have the child layout controls that also live under Dimensions. My weakly held opinion is that it'd be better to keep aspect ratio with all these controls for now and look at how they all sit together as a separate task from this block support. Here's how the Aspect Ratio control looks sitting alongside child layout controls, too. There sure are a lot of controls now!

<img width="1276" alt="image" src="https://github.com/WordPress/gutenberg/assets/14988353/091240e9-5c6e-4fcc-80e9-1c208df71baf">

---

> In the Cover block, when I set a minimum height and then an aspect ratio, the cover no longer obeys the minimum height. I'd expect it to reset the minimum height when I do this, which is what happens in the Group block.

This one is slightly nuanced — this block support shouldn't need to know about the Cover block's ad hoc controls / values for minimum height, so if we want to clear out the Cover block's value, we'll need to figure out an elegant way to do it. I was hoping we could look at that in a follow-up potentially as it's a bit more related to the Cover block than the overall block support. Let me know if you think it's a blocker.

> It makes sense to me that you can control the default Aspect ratio for all Cover blocks using Global Styles, but I'm not sure about Group? Group is heavily used by block themes for layouts and so setting it nearly always makes the entire site look broken.

Agreed. While you were reviewing I pushed a commit that hides the Group block from global styles for now in https://github.com/WordPress/gutenberg/pull/56897/commits/39873bf9e2ec0e32194d94b76b5e5e7cc0550bd6

> The aspect ratio dropdown displays Original as the selected item both when there is no aspect ratio set on the block and when the aspect ratio is explicitly set to auto. I find this a bit confusing—it made me think that global styles inheritance wasn't working.

Yes, that one's a little tricky, too. This PR re-uses the existing behaviour from the `AspectRatioTool` component. I lean a bit toward seeing if we can tweak this in a follow-up so that we can look at that component in a bit more detail. In some ways it's a similar challenge to the block spacing controls where the unset position is the same visually as if it's set to `0` 🤔

> I found it a bit odd that Featured Image has an Aspect ratio control but that it a) didn't display the same values as Cover and Group; and b) wasn't adjustable using Global Styles.

Yes, just to add to what Rich mentioned, if and when this block support lands, it'd be good to see which blocks could potentially be moved over to using the block support, that way they could get global styles support, too.

---

Thanks again for the reviews and feedback everyone! To recap it looks like:

* We don't yet have consensus on where the aspect ratio controls should live: it'd be good to know if it's a blocker for this feature, or if we can land it in the dimensions panel and explore alternatives in follow-ups.
* Rob flagged some bits of friction that would be good to tidy up. I'm leaning toward seeing if we could fix those things in follow-ups as it might be a bit easier to do that iteratively since this PR is pretty big now.
* We all seem to agree that hiding the Group block from getting aspect ratio in global styles is a good thing (done in https://github.com/WordPress/gutenberg/pull/56897/commits/39873bf9e2ec0e32194d94b76b5e5e7cc0550bd6)

Please let me know which issues folks think are blockers to landing, and / or if I've missed anything! I imagine if this feature is to make it in for WP 6.5, we'd probably want it to be included in Gutenberg 17.6 so that it can be tested in the plugin prior to being put in a core release 🤔

tellthemachines reviewed (approved):
Thanks for the iteration! I think this is in a good place to go into the plugin. 

I see the question of where the control should live as part of a wider conceptual problem that affects most of the controls currently in Dimensions (and some of the ad-hoc controls added to blocks like Image too) and that needs to be explored as its own problem. I don't think that should block this feature from being added.

Relatedly, if possible it would be good to change some of those ad-hoc controls to use the proper block support, but that should definitely be a follow-up task 😅

andrewserong commented:
Thanks for confirming and for the approving review @tellthemachines! 🙇

@jasmussen and @richtabor — please let me know if any of the concerns discussed thus far are blockers for either of you. Unless anyone objects, I'll aim to merge this PR in sometime tomorrow (just to give everyone a chance to catch up on the discussion). Then, happy to continue tweaking in follow-ups.

jasmussen commented:
I'd love to land it, and from my quick testing, this is working solidly. I think Rich makes a solid point that we do want to consider putting this control in a place that is unified across image, cover, and featured image blocks, and that could be a new panel. I don't know that needs to block this from landing, in fact it may be good to merge and backport this, it'll be a really neat 6.5 feature.

One other question to consider in followups: is "Original" a good term for when there's no custom aspect ratio supplied? For images and even videos, it's correct since there are intrinsic dimensions in the media. But for featured image and cover, technically there are no intrinsic dimensions. It's a bit meta, but I wonder if "Auto" would work across all instead of "Original"?

andrewserong commented:
Thanks folks! I'll merge this in now and then will create an issue for each of the bits of friction to be addressed in follow-ups. Thanks again for all the reviews and feedback! 🙇

andrewserong commented:
Just adding the Backport to Gutenberg RC label to see if we can get this in for GB 17.6.

andrewserong commented:
I've opened up a separate issue for the tasks and issues discussed here over in: https://github.com/WordPress/gutenberg/issues/58230 — feel free to add anything I might have missed, and we can continue the discussion over there, too 🙂

c4rl0sbr4v0 commented:
I just cherry-picked this PR to the release/17.6 branch to get it included in the next release: https://github.com/WordPress/gutenberg/commit/bcc44acabb2d4aef2a0571568ac92a09589babd0

youknowriad commented:
Let's add this to the #57959 issue as it requires some manual php backports and please consider opening a core trac ticket and backport. Thank you.

richtabor commented:
> We all seem to agree that hiding the Group block from getting aspect ratio in global styles is a good thing (done in https://github.com/WordPress/gutenberg/commit/39873bf9e2ec0e32194d94b76b5e5e7cc0550bd6)

I'm noticing the aspect ratio support still on the group block:

![CleanShot 2024-01-25 at 08 30 31](https://github.com/WordPress/gutenberg/assets/1813435/b19ecc0f-627f-4840-9730-4150e424c8c6)

djcowan commented:
G. O. L. D.

getdave commented:
I noticed this PR was merged after I raised the [PHP Sync Tracking Issue for WP 6.5](https://github.com/WordPress/gutenberg/issues/57959) and has changed PHP files that _may_ need backporting to WP Core.

I've now added it to the Tracking Issue.

andrewserong commented:
Thanks folks, I'll work on backports this week.

> I'm noticing the aspect ratio support still on the group block:

@richtabor it's only hidden at the global styles level for the Group block. At the individual block level it's supported as it's still useful there. The main thing we wanted to avoid was folks accidentally setting all Group blocks to a particular aspect ratio and then breaking site layouts.

andrewserong commented:
> I noticed this PR was merged after I raised the https://github.com/WordPress/gutenberg/issues/57959 and has changed PHP files that may need backporting to WP Core.

Update: I've opened [a trac ticket](https://core.trac.wordpress.org/ticket/60365) and a backport PR over here: https://github.com/WordPress/wordpress-develop/pull/5963, and updated the backports tracking issue with links to those two 🙂

richtabor commented:
> @richtabor it's only hidden at the global styles level for the Group block. At the individual block level it's supported as it's still useful there

How is aspect ratio on a group block helpful? If we don't have the cover block's position controls, it feels unresponsive if the group block has more than a couple blocks, or is fullwidth. I think there's too many variables where a group block is used differently from a cover block, where it doesn't make as much sense to include.  

https://github.com/WordPress/gutenberg/assets/1813435/a8c67020-12e8-4900-8fc3-676697e3726f

andrewserong commented:
> How is aspect ratio on a group block helpful?

That's a good point. It's hard to use with the Group block as it only really feels at home with a Row or Stack block where the content is small enough for it to be useful and where we have the justification controls. That makes it a bit of a niche feature. Here's an example of using it with the Row variation:

https://github.com/WordPress/gutenberg/assets/14988353/8d98abb3-6868-49d4-804f-b576da2dd574

In its current form, I agree the aspect ratio controls are likely too prominent on the Group block for something that'll likely be used very rarely. I'm in favour of removing support on the Group block for now, and leaving it just on the Cover block for WP 6.5. We can then revisit for subsequent releases if we think it feels missing not having it on the Group block.

I have a PR up to remove support on the Group block over in #58414

andrewserong commented:
Update: #58414 has been merged, so support is only for the Cover block for now.

paaljoachim commented:
As a result of adding the aspect ratio to the Cover block...
This makes it possible for WooCommerce to switch the Featured Image block (Product image block) to use the Cover block. It will keep the product images the same size. https://github.com/woocommerce/woocommerce/issues/44387

andrewserong commented:
Draft dev note:

---

## Aspect ratio block support

In WordPress 6.5, a new aspect ratio block support has been added, with the Cover block opted-in by default. For themes using the `appearanceTools` feature in `theme.json`, the control will be available in the inspector controls, under Dimensions.

The feature allows users to set an aspect ratio for the Cover block, and is mutually exclusive with `min-height` rules. If an aspect ratio is set, then `min-height` is `unset` to ensure that height rules do not conflict with the aspect ratio. The aspect ratio rules will be output at render-time for the block via an inline style applied to the block's wrapper.

Note that themes or blocks that add width or height rules to a block will need to take care in testing compatibility with `aspect-ratio` if they are to opt-in to aspect ratio support. Aspect ratio tends to work most flexibly when width or height rules are not set.

### How to add aspectRatio support to a theme

There are two ways to add support for `aspectRatio` to a block theme. The simplest is to opt in to the `appearanceTools` setting, which automatically enables a number of design tools ([read more in the developer handbook](https://developer.wordpress.org/block-editor/how-to-guides/themes/theme-json/#opt-in-into-ui-controls)).

For themes that wish to have more granular control over which UI tools are enabled, the `aspectRatio` support can be opted into by setting `settings.dimensions.aspectRatio` to `true` in `theme.json`. For example:

```
{
	"settings": {
		"dimensions": {
			"aspectRatio": true
```
